pkg/machines/helpers.js

```diff
@@ -375,3 +377,164 @@ export function timeoutedPromise(promise, delay, afterTimeoutHandler, afterTimeo
+    let nodeDev;
+    switch (hostdev.type) {
+    case "usb": {
+        const vendorId = hostdev.source.vendor.id;
+        const productId = hostdev.source.product.id;
+        nodeDev = nodeDevices.find(d => {
+            if (vendorId &&
+                productId &&
+                d.capability.vendor &&
+                d.capability.product &&
+                d.capability.vendor.id == vendorId &&
+                d.capability.product.id == productId)
+                return true;
+        });
+        break;
+    }
+    case "pci": {
+        // convert hexadecimal number in string to decimal number in string
+        const domain = parseInt(hostdev.source.address.domain, 16).toString();
+        const bus = parseInt(hostdev.source.address.bus, 16).toString();
+        const slot = parseInt(hostdev.source.address.slot, 16).toString();
+        const func = parseInt(hostdev.source.address.func, 16).toString();
+        nodeDev = nodeDevices.find(d => {
+            if ((domain && bus && slot && func) &&
+                d.capability.domain &&
+                d.capability.bus &&
+                d.capability.slot &&
+                d.capability.function &&
+                d.capability.domain._value == domain &&
+                d.capability.bus._value == bus &&
+                d.capability.slot._value == slot &&
+                d.capability.function._value == func)
+                return true;
+        });
+        break;
+    }
+    case "scsi": {
+        const bus = hostdev.source.address.bus;
+        const target = hostdev.source.address.target;
+        const unit = hostdev.source.address.unit;
+        nodeDev = nodeDevices.find(d => {
+            if ((bus && target && unit) &&
+                d.capability.bus &&
+                d.capability.lun &&
+                d.capability.target &&
+                d.capability.bus._value == bus &&
+                d.capability.lun._value == unit &&
+                d.capability.target._value == target)
+                return true;
+        });
+        break;
+    }
+    case "scsi_host": {
+        // TODO add scsi_host
+        nodeDev = undefined;
+        break;
+    }
+    case "mdev": {
+        const uuid = hostdev.source.address.uuid;
+        nodeDev = nodeDevices.find(d => {
+            if (d.path &&
+                d.path._value.contains(uuid))
+                return true;
+        });
+        break;
+    }
+    }
+    return nodeDev;
+}
+
+export function getBootOrderDevices(vm) {
+    let devices = [];
+    let diskCount = 0;
+    let ifaceCount = 0;
+
+    // Some disks and interfaces may have boot order in os->boot (legacy)
+    if (vm.osBoot) {
+        for (let i = 0; i < vm.osBoot.length; i++) {
+            const boot = vm.osBoot[i];
+
+            if (boot.type === "disk" | boot.type === "cdrom" | boot.type === "fd") {
+                devices.push({
+                    device: Object.values(vm.disks)[diskCount++],
```

Comment:
As I mentioned below, I do not skip first disk only, but it can skip multiple disks. And in this line of code, diskCount++ doesn't do any skipping, it's there to really just keep count. The skipping of disks comes later when the cycle for cycling disks starts from diskCount value.

It seems that libvirt itself sorts the disks in devices according to target.

Reply:
> As I mentioned below, I do not skip first disk only, but it can skip multiple disks. And in this line of code, diskCount++ doesn't do any skipping, it's there to really just keep count. The skipping of disks comes later when the cycle for cycling disks starts from diskCount value.
> 

I understand it. What I said it that skipping might not be correct. Let's sat that have have :+1:   <boot dev='cdrom'/> in the os element.

Which means that the first chrom disk should be bootable. And let's asume that there is not other <boot> element. So diskCount == 1.

Then you do to devices section in XML, and let's say it's like:
``` xml
  <disk type='volume' device='disk'>
    <driver name='qemu' type='raw'/>
    <source pool='blk-pool0' volume='blk-pool0-vol0'/>
    <target dev='vda' bus='virtio'/>
  </disk>
  <disk type='block' device='cdrom'>
    <driver name='qemu' type='raw'/>
    <target dev='hdd' bus='ide' tray='open'/>
    <readonly/>
  </disk>
```
What your code does is mark the first disk (which is the vda) as bootable but it should not do such.. Correct me if I am wrong.

> It seems that libvirt itself sorts the disks in devices according to target.

Reply:
Oh right, I misunderstood the problem here. Alright, so I kinda changed the approach. Instead, all disks and interfaces are copied into arrays.
If XML uses the old way for boot order (os->boot), we find corresponding device in those arrays, and then remove it from the array. So there are only devices without boot order left.